### PR TITLE
gobject-introspection: clean-ups & build docs

### PR DESCRIPTION
**Pull request by jtojnar**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
@GrahamcOfBorg build gobject-introspection

**Comment by jtojnar** [comment]
@GrahamcOfBorg build gobject-introspection

File: pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch [hunk 1]
```diff
@@ -140,3 +140,46 @@
...
  def extract_libtool(la_file):
+--- a/tests/scanner/test_shlibs.py
++++ b/tests/scanner/test_shlibs.py
+@@ -7,6 +7,30 @@ from giscanner.shlibs import resolve_from_ldd_output, sanitize_shlib_path
+ 
+ class TestLddParser(unittest.TestCase):
+ 
++    def test_resolve_from_ldd_output_nix(self):
++        output = '''\
++            libglib-2.0.so.0 => @nixStoreDir@/gmrf09y7sfxrr0mcx90dba7w41jj2kzk-glib-2.58.1/lib/libglib-2.0.so.0 (0x00007f0ee1b28000)
```
Comment (Mic92) [inline pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch:151]: I don't think the part after `@nixStoreDir@` is stable or consistent across different architectures.

File: pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch [hunk 2]
```diff
@@ -140,3 +140,46 @@
...
++            libuuid.so.1 => @nixStoreDir@/q0kgnq21j0l2yd77gdlld371246cwghh-util-linux-2.32.1/lib/libuuid.so.1 (0x00007f0edcec5000)
++            librt.so.1 => @nixStoreDir@/g2yk54hifqlsjiha3szr4q3ccmdzyrdv-glibc-2.27/lib/librt.so.1 (0x00007f0edccbd000)
++            libstdc++.so.6 => @nixStoreDir@/3v5r7fkrbkw2qajadvjbf6p6qriz9p1i-gcc-7.3.0-lib/lib/libstdc++.so.6 (0x00007f0edc936000)
++            libgcc_s.so.1 => @nixStoreDir@/g2yk54hifqlsjiha3szr4q3ccmdzyrdv-glibc-2.27/lib/libgcc_s.so.1 (0x00007f0edc720000)
++        '''
++        libraries = ['glib-2.0', 'gio-2.0']
++
++        self.assertEqual(
++            ['@nixStoreDir@/gmrf09y7sfxrr0mcx90dba7w41jj2kzk-glib-2.58.1/lib/libglib-2.0.so.0',
++             '@nixStoreDir@/gmrf09y7sfxrr0mcx90dba7w41jj2kzk-glib-2.58.1/lib/libgio-2.0.so.0'],
```
Comment (Mic92) [inline pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch:170]: Ah. It's just a test.

File: pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch (hunk 1 above)
Comment (jtojnar) [reply pkgs/development/libraries/gobject-introspection/absolute_shlib_path.patch:151]: That is just a random string for a test of pattern matching.

File: pkgs/development/libraries/gobject-introspection/default.nix [hunk 3]
```diff
@@ -1,61 +1,94 @@
...
+  ];
+
+  buildInputs = [
+    flex
+    bison
+    python3
+    setupHook /*move .gir*/
+  ] ++ stdenv.lib.optionals stdenv.isDarwin [
+    cctools
+  ];
```
Comment (Mic92) [inline pkgs/development/libraries/gobject-introspection/default.nix:75]: I would say that all of these tools should be in `nativeBuildInputs`.

Comment (jtojnar) [reply pkgs/development/libraries/gobject-introspection/default.nix:75]: Thanks for pointing this out.

File: pkgs/development/libraries/gobject-introspection/default.nix [hunk 4]
```diff
@@ -59,19 +59,22 @@ stdenv.mkDerivation rec {
...
+    docbook_xml_dtd_43 # FIXME: remove in next release
     docbook_xml_dtd_45
+    python3
+    setupHook # move .gir files
   ];
 
   buildInputs = [
-    flex
-    bison
     python3
```
Comment (Mic92) [inline pkgs/development/libraries/gobject-introspection/default.nix:73]: cc @FRidh from a cross-compiling perspective we usually used the build-native python interpreter to build python packages, no?

File: pkgs/development/libraries/gobject-introspection/default.nix [hunk 5]
```diff
@@ -59,19 +59,22 @@ stdenv.mkDerivation rec {
...
     ninja
     pkg-config
+    flex
+    bison
     gtk-doc
     docbook-xsl-nons
-    docbook_xml_dtd_43
+    docbook_xml_dtd_43 # FIXME: remove in next release
     docbook_xml_dtd_45
+    python3
```
Comment (Mic92) [inline pkgs/development/libraries/gobject-introspection/default.nix:68]: Ah. It's in both. That would be fine I suppose.

**Comment by jtojnar** [comment]
@GrahamcOfBorg build gobject-introspection

File: pkgs/development/libraries/gobject-introspection/default.nix (hunk 4 above)
Comment (FRidh) [reply pkgs/development/libraries/gobject-introspection/default.nix:73]: yes, the build-native one is used in `buildPythonPackage`.

Comment (FRidh) [reply pkgs/development/libraries/gobject-introspection/default.nix:73]: maybe they link against the interpreter?

Comment (jtojnar) [reply pkgs/development/libraries/gobject-introspection/default.nix:73]: Build scripts need build python and some gobject utils are written in python so they need host python.

**Comment by Ericson2314** [comment]
`liblibarchive` builds now

**Comment by jtojnar** [comment]
@GrahamcOfBorg build gobject-introspection

**Comment by jtojnar** [comment]
@GrahamcOfBorg eval

**Comment by Ericson2314** [comment]
I just built x86_64 linux and darwin.

**Pull request merged**